File: app/spring/src/main/java/az/ecommerce/payment/domain/event/PaymentFailedEvent.java
package az.ecommerce.payment.domain.event;

import az.ecommerce.payment.domain.valueobject.PaymentId;
import az.ecommerce.shared.domain.DomainEvent;

import java.time.Instant;
import java.util.UUID;

public record PaymentFailedEvent(UUID eventId, Instant occurredAt, PaymentId paymentId, UUID orderId, String reason) implements DomainEvent {
    public static PaymentFailedEvent of(PaymentId id, UUID orderId, String reason) {
        return new PaymentFailedEvent(UUID.randomUUID(), Instant.now(), id, orderId, reason);
    }
}

File: app/laravel/src/Payment/Domain/Events/PaymentFailedEvent.php
<?php

declare(strict_types=1);

namespace Src\Payment\Domain\Events;

use Src\Shared\Domain\DomainEvent;

/**
 * PAYMENT FAILED EVENT (Domain Event)
 * ====================================
 * Ödəniş uğursuz olduqda baş verən hadisə.
 *
 * Ödəniş nəyə görə uğursuz ola bilər?
 * - Kredit kartının balansı çatmır
 * - Kartın müddəti bitib
 * - Gateway xətası (texniki problem)
 * - Fraud detection (saxtakarlıq aşkarlanması)
 *
 * Bu event-dən sonra:
 * - Müştəriyə "ödəniş uğursuz oldu" bildirişi göndərilə bilər
 * - PaymentFailedIntegrationEvent ilə Order moduluna məlumat verilər
 */
final class PaymentFailedEvent extends DomainEvent
{
    public function __construct(
        private readonly string $paymentId,
        private readonly string $orderId,
        private readonly string $reason,
    ) {
        parent::__construct();
    }

    public function paymentId(): string
    {
        return $this->paymentId;
    }

    public function orderId(): string
    {
        return $this->orderId;
    }

    public function reason(): string
    {
        return $this->reason;
    }

    public function eventName(): string
    {
        return 'payment.failed';
    }

    public function toArray(): array
    {
        return [
            'payment_id' => $this->paymentId,
            'order_id' => $this->orderId,
            'reason' => $this->reason,
        ];
    }
}
